include/ignition/common/TempDirectory.hh

```diff
@@ -0,0 +1,106 @@
+    /// \brief Return the path to a directory suitable for temporary files.
+    ///
+    /// Calls std::filesystem::temp_directory_path, refer to the standard
+    /// documentation for your platform for behaviors.
+    /// \return A directory suitable for temporary files.
+    std::string IGNITION_COMMON_VISIBLE tempDirectoryPath();
+
+    /// \brief Create a directory in the tempDirectoryPath by expanding
+    /// a name template
+    ///
+    /// On execution, will create the directory:
+    ///  "_parentPath"/"_baseName" + "XXXXXX", where XXXXXX will be filled
+    /// out by an OS-appropriate method (eg mkdtmp/_mktemp_s)
+    ///
+    /// \param[in] _baseName String to be prepended to the expanded template
+    /// \param[in] _parentPath Location to create the directory
+    /// \param[in] _warningOp Allow or suppress filesystem warnings
+    /// \return Path to newly-created temporary directory
+    std::string IGNITION_COMMON_VISIBLE createTempDirectory(
+        const std::string &_baseName,
+        const std::string &_parentPath,
+        const FilesystemWarningOp _warningOp = FSWO_LOG_WARNINGS);
+
+    /// \class TempDirectory TempDirectory.hh ignitin/common/TempDirectory.hh
+    /// \brief Create a temporary directory in the OS temp location.
+    class IGNITION_COMMON_VISIBLE TempDirectory
+    {
+      /// \brief Create a directory in the tempDirectoryPath by expanding
+      /// a name template.  This directory can also be automatically cleaned
+      /// up when the object goes out of scope.
+      ///
+      /// The TempDirectory will have the form $TMPDIR/_subdir/_prefixXXXXX/
+      ///
+      /// \param[in] _prefix String to be expanded for the template
+      /// \param[in] _subDir Subdirectory in OS $TMPDIR, if desired
+      /// \param[in] _cleanup True to indicate that the filesystem should
+      ///   be cleaned as part of the destructor
+      public: TempDirectory(const std::string &_prefix = "temp_dir",
+                            const std::string &_subDir = "ignition",
+                            bool _cleanup = true);
+
+      /// \brief Destroy the temporary directory, removing from filesystem
+      /// if cleanup is true.
+      public: ~TempDirectory();
+
+      /// \brief Indicate if the TempDirectory object is in a valid state
+      /// and that the folder exists on the filesystem
+      /// \return true if the TempDirectory is valid
+      public: bool Valid() const;
+
+      /// \brief Set if the folder on disk should be cleaned.
+      ///
+      /// This is useful if you wish to clean by default during a test, but
+      /// retain the contents of the TempDirectory if the test fails.
+      /// \param[in] _cleanup True to indicate that the filesystem should
+      ///  be cleaned as part of the destructor
+      public: void Cleanup(bool _cleanup);
+
+      /// \brief Retrieve the current cleanup flag state
+      /// \return true if filesystem cleanup will occur
+      public: bool Cleanup() const;
```

Comment:
These functions' names could be misleading, sounding like the cleanup happens when they're called

Reply:
I changed them to `DoCleanup`, which mirrors what the internal member is also called.